### PR TITLE
[inotify] Add support for the IN_CLOSE_WRITE event

### LINKED ISSUE #93: Modifed event triggered twice
if I copy and paste a file on OSX lion
it triggered created event once and modified event twice
should it be just trigger once modified event?

`
INFO:root:Created file: /Users/ouyangjichao/BoxSync/rename/testwatchdog.py
INFO:root:Modified file: /Users/ouyangjichao/BoxSync/rename/testwatchdog.py
INFO:root:Modified file: /Users/ouyangjichao/BoxSync/rename/testwatchdog.py
`

### PR DESCRIPTION
**Pull request by ysard**
Hi, after testing #690 I found that the event was still not captured by the "inotify.InotifyObserver" observer. Here are my additions to this PR that got me on track. This should also close #93.

- "Full" implementation of the capture of IN_CLOSE_NOWRITE & IN_CLOSE_WRITE for the detection of the closing files (only).
- Unit tests

Please note that IN_CLOSE_NOWRITE is triggered very often with the IN_ISDIR mask; those events are filtered, see:

https://github.com/ysard/watchdog/blob/0de211c9864bce310a8fd9e5634e6a3e1ea3f9bc/src/watchdog/observers/inotify.py#L178-L182

Some additional events (IN_ACCESS, IN_OPEN) could easily be added in the future.
I don't know how far you want to go in the inotify support because it implies to make the lib less portable so I don't add anything more.

Waiting for your review :)

### CODE REVIEW

**Comment by BoboTiG** [comment]
Great!

Could you rebase your branch to fix conflicts?

I dis not look, but is ot necessary for IN_CLOSE_NOWRITE to generate an event? If ghe file was not modified, then the event should not be sent. WDYT?

**Comment by ysard** [comment]
It's now synced with master. I also fixed a test that failed in Travis.

About the other point, it's a good question...

- For the sake of completeness and if we stick to the strict definition of the event: these are the 2 masks that must be considered.
Moreover this is already provided here (in helper user-space events):
https://github.com/gorakhargosh/watchdog/blob/847a059ae717424baea81a949c7726fc3520bb8a/src/watchdog/observers/inotify_c.py#L93


- It's hard to know if people also expect notification when a file closes after read-only;  from what I read in the issues and on stackoverflow it's `IN_CLOSE_WRITE` that's currently the most lacking.

In practice, the use of IN_CLOSE_NOWRITE is probably uncommon and "can" generate a higher processing cost because the slightest file traversal by a program emits `IN_CLOSE_NOWRITE | IN_ISDIR`.

You will always find someone who needs an unimplemented event for a project...


The safest thing to do would be to let the user choose the mask he is interested in; however the Inotify class (and the keyword argument `event_mask` of its constructor) is not accessible from the API.

---

Proof of the overhead with `IN_CLOSE_WRITE | IN_CLOSE_NOWRITE`:

```
fd = open("./dir/hello", "w")
<InotifyEvent: src_path=b'./dir/hello', wd=8, mask=IN_CREATE, cookie=0, name=hello>
<InotifyEvent: src_path=b'./dir', wd=1, mask=IN_CLOSE_NOWRITE|IN_ISDIR, cookie=0, name=dir>
<InotifyEvent: src_path=b'./dir', wd=8, mask=IN_CLOSE_NOWRITE|IN_ISDIR, cookie=0, name=>
<InotifyEvent: src_path=b'./dir/hello', wd=8, mask=IN_CLOSE_NOWRITE, cookie=0, name=hello>

fd.write("a")
fd.flush()
<InotifyEvent: src_path=b'./dir/hello', wd=8, mask=IN_MODIFY, cookie=0, name=hello>
<InotifyEvent: src_path=b'./dir/hello', wd=8, mask=IN_CLOSE_NOWRITE, cookie=0, name=hello>

fd.close()
<InotifyEvent: src_path=b'./dir/hello', wd=8, mask=IN_CLOSE_WRITE, cookie=0, name=hello>


os.remove("./dir/hello")

<InotifyEvent: src_path=b'./dir/hello', wd=8, mask=IN_DELETE, cookie=0, name=hello>
<InotifyEvent: src_path=b'./dir', wd=1, mask=IN_CLOSE_NOWRITE|IN_ISDIR, cookie=0, name=dir>
<InotifyEvent: src_path=b'./dir', wd=8, mask=IN_CLOSE_NOWRITE|IN_ISDIR, cookie=0, name=>
```

With only IN_CLOSE_WRITE:
```
<InotifyEvent: src_path=b'./dir/hello', wd=8, mask=IN_CREATE, cookie=0, name=hello>
<InotifyEvent: src_path=b'./dir/hello', wd=8, mask=IN_MODIFY, cookie=0, name=hello>
<InotifyEvent: src_path=b'./dir/hello', wd=8, mask=IN_CLOSE_WRITE, cookie=0, name=hello>
```

---

Other note: As indicated in the doc, the observers are specific to each OS (https://pythonhosted.org/watchdog/api.html#watchdog.observers.Observer).

However, to give access to the event we are interested in here, it was necessary to modify the class `watchdog.events.FileSystemEventHandler` by adding the `on_closed` method.

This class is meant to be universal and this modification is not portable. As the issue #217 reminds us, we must warn the user that this method is only available under GNU/Linux.

=> Should I add a note in the docstring?

The current issue is a long-standing problem (#245, #184, #313, #280, even #119 for the IN_ACCESS event).
It is still possible for the user to achieve this by adding the desired methods if the lib does not offer them via monkey patching, but this is not ideal.

File: src/watchdog/events.py [hunk 1]
```diff
@@ -212,6 +217,19 @@ class FileMovedEvent(FileSystemMovedEvent):
...
+
+    event_type = EVENT_TYPE_CLOSED
+
+    def __init__(self, src_path):
+        super(FileClosedEvent, self).__init__(src_path)
+
+    def __repr__(self):
+        return ("<%(class_name)s: src_path=%(src_path)r>"
+                ) % (dict(class_name=self.__class__.__name__,
+                          src_path=self.src_path))
```
Comment (BoboTiG) [inline src/watchdog/events.py:231]: Not needed. See previous above.

**Comment by BoboTiG** [comment]
IMO the `IN_CLOSE_NOWRITE` event will add too much noise. And the benefits of having the `IN_CLOSE` event is to be able to know when a file modification has been done for real.

Could you do that?
You could comment the code you added to support `IN_CLOSE_NOWRITE` to keep the track for later. And duplicate and adapt https://github.com/gorakhargosh/watchdog/blob/847a059ae717424baea81a949c7726fc3520bb8a/src/watchdog/observers/inotify_c.py#L93 to remove `IN_CLOSE_NOWRITE`.

**Comment by BoboTiG** [comment]
Could you add a line in the logs (+ your GitHub nickename and @lukassup to our beloved contributors):
```rst
- [inotify] Add support for ``IN_CLOSE_WRITE`` events. A ``FileCloseEvent`` event will be fired. Note that ``IN_CLOSE_NOWRITE`` is not handled to prevent much noise. (`#184 <https://github.com/gorakhargosh/watchdog/pull/184>`_, `#245 <https://github.com/gorakhargosh/watchdog/pull/245>`_, `#280 <https://github.com/gorakhargosh/watchdog/pull/280>`_, `#313 <https://github.com/gorakhargosh/watchdog/pull/313>`_, `#690 <https://github.com/gorakhargosh/watchdog/pull/690>`_)
```

File: src/watchdog/observers/inotify.py [hunk 2]
```diff
@@ -170,6 +171,13 @@ def queue_events(self, timeout, full_events=False):
                 cls = DirCreatedEvent if event.is_directory else FileCreatedEvent
                 self.queue_event(cls(src_path))
                 self.queue_event(DirModifiedEvent(os.path.dirname(src_path)))
+            elif event.is_close_write and not event.is_directory:
+                cls = FileClosedEvent
+                self.queue_event(cls(src_path))
+                self.queue_event(DirModifiedEvent(os.path.dirname(src_path)))
+            elif event.is_close_nowrite and not event.is_directory:
+                cls = FileClosedEvent
+                self.queue_event(cls(src_path))
```
Comment (BoboTiG) [inline src/watchdog/observers/inotify.py:180]:
```suggestion
            # elif event.is_close_nowrite and not event.is_directory:
            #     cls = FileClosedEvent
            #     self.queue_event(cls(src_path))
```

Comment (ysard) [reply src/watchdog/observers/inotify.py:180]: Yeah ok I will rebase 66c1e00 to have atomic changes about the removing of this event

**Comment by ysard** [comment]
It's done now ;)

**Review by BoboTiG — APPROVED** [review]

**Comment by BoboTiG** [comment]
Thanks a lot @ysard 💪

**Comment by Yardentexel** [comment]
is this implemented only for unix?

**Comment by BoboTiG** [comment]
Yes, `IN_CLOSE_WRITE` is Unix-only.

**Comment by Yardentexel** [comment]
the answer is yes :)

tried in windows - on_close not working
tried from WSL2 (windows sub system) - on_close not working
tried unix docker image -> works 👍

**Comment by UnitedMarsupials** [comment]
> tried unix docker image 

You meant _Linux_ image, didn't you? It will not work for other Unixes -- as the `kqueue` support is not implemented, for example.

**Comment by BoboTiG** [comment]
> Why it was not added to LoggingEventHandler? ... crying_cat_face

I finally heard you: f3dfc4be0c6ec36c2db6b4afa075004e71adf87e :)

You should have open an issue, I didn't catch that until now 😅

**Pull request merged**